File: mojo/public/python/mojo_application/service_provider_impl.py
# Copyright 2015 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

"""Python implementation of the ServiceProvider interface."""

import logging

import service_provider_mojom

class ServiceProviderImpl(service_provider_mojom.ServiceProvider):
  def __init__(self, provider):
    self._provider = provider
    self._name_to_service_connector = {}

  def AddService(self, service_class):
    self._name_to_service_connector[service_class.manager.name] = service_class

  def ConnectToService(self, interface_name, pipe):
    if interface_name in self._name_to_service_connector:
      service = self._name_to_service_connector[interface_name]
      service.manager.Bind(service(), pipe)
    else:
      logging.error("Unable to find service " + interface_name)

File: sky/shell/platform/android/io/flutter/view/ServiceProviderImpl.java
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.view;

import android.content.Context;
import android.util.Log;
import java.util.HashSet;

import org.chromium.mojo.bindings.ConnectionErrorHandler;
import org.chromium.mojo.bindings.Interface.Binding;
import org.chromium.mojo.system.Core;
import org.chromium.mojo.system.MessagePipeHandle;
import org.chromium.mojo.system.MojoException;
import org.chromium.mojo.system.impl.CoreImpl;
import org.chromium.mojom.mojo.ServiceProvider;

/**
 * A collection of services.
 **/
class ServiceProviderImpl implements ServiceProvider {
    private static final String TAG = "ServiceProviderImpl";

    private Core mCore;
    private FlutterView mView;
    private ServiceRegistry mRegistry;
    private HashSet<Binding> mBindings = new HashSet<Binding>();

    ServiceProviderImpl(Core core, FlutterView view, ServiceRegistry registry) {
        assert core != null;
        assert view != null;
        mCore = core;
        mView = view;
        mRegistry = registry;
    }

    @Override
    public void close() {}

    @Override
    public void onConnectionError(MojoException e) {}

    @Override
    public void connectToService(String interfaceName, MessagePipeHandle pipe) {
        ServiceFactory factory = mRegistry.get(interfaceName);
        if (factory == null) {
            pipe.close();
            return;
        }
        final Binding binding = factory.connectToService(mView, mCore, pipe);
        mBindings.add(binding);
        binding.registerErrorHandler(new ConnectionErrorHandler() {
            @Override
            public void onConnectionError(MojoException e) {
                Log.w(TAG, "Flutter service provider connection error", e);
                mBindings.remove(binding);
            }
        });
    }

    public void unbindServices() {
        for (Binding binding : mBindings) {
            binding.unbind().close();
        }
        mBindings.clear();
    }
}
